src/constraints.rs

```diff
@@ -0,0 +1,536 @@
+use crate::{Groth16, PreparedVerifyingKey, Proof, VerifyingKey};
+use ark_crypto_primitives::snark::constraints::{CircuitSpecificSetupSNARKGadget, SNARKGadget};
+use ark_crypto_primitives::snark::{BooleanInputVar, SNARK};
+use ark_ec::{AffineCurve, PairingEngine};
+use ark_r1cs_std::groups::CurveVar;
+use ark_r1cs_std::{
+    alloc::{AllocVar, AllocationMode},
+    bits::boolean::Boolean,
+    bits::uint8::UInt8,
+    eq::EqGadget,
+    pairing::PairingVar,
+    ToBitsGadget, ToBytesGadget,
+};
+use ark_relations::r1cs::{Namespace, SynthesisError};
+use ark_std::{borrow::Borrow, marker::PhantomData, vec::Vec};
+
+/// The proof variable for the Groth16 construction
+#[derive(Derivative)]
+#[derivative(Clone(bound = "P::G1Var: Clone, P::G2Var: Clone"))]
+pub struct ProofVar<E: PairingEngine, P: PairingVar<E>> {
+    /// The `A` element in `G1`.
+    pub a: P::G1Var,
+    /// The `B` element in `G2`.
+    pub b: P::G2Var,
+    /// The `C` element in `G1`.
+    pub c: P::G1Var,
+}
+
+/// The verifying key variable for the Groth16 construction
+#[derive(Derivative)]
+#[derivative(
+    Clone(bound = "P::G1Var: Clone, P::GTVar: Clone, P::G1PreparedVar: Clone, \
+    P::G2PreparedVar: Clone, ")
+)]
+pub struct VerifyingKeyVar<E: PairingEngine, P: PairingVar<E>> {
+    /// The `alpha * G`, where `G` is the generator of `E::G1`.
+    pub alpha_g1: P::G1Var,
+    /// The `beta * G`, where `G` is the generator of `E::G2`.
+    pub beta_g2: P::G2Var,
+    /// The `gamma * G`, where `G` is the generator of `E::G2`.
+    pub gamma_g2: P::G2Var,
+    /// The `delta * G`, where `G` is the generator of `E::G2`.
+    pub delta_g2: P::G2Var,
+    /// The `gamma^{-1} * (alpha * a_i + beta * b_i + c_i) * H`, where `H` is the generator of `E::G1`.
+    pub gamma_abc_g1: Vec<P::G1Var>,
```

Comment:
The existing comments don't really convey any extra information.

```suggestion
    #[doc(hidden)]
    pub alpha_g1: P::G1Var,
    #[doc(hidden)]
    pub beta_g2: P::G2Var,
    #[doc(hidden)]
    pub gamma_g2: P::G2Var,
    #[doc(hidden)]
    pub delta_g2: P::G2Var,
    #[doc(hidden)]
    pub gamma_abc_g1: Vec<P::G1Var>,
```